### ISSUE TITLE
ARCHITECTURE: Migrate to TypeScript

### ISSUE THREAD

**Issue by arnog** [arnog/mathlive#418]
TypeScript is a strict superset of JavaScript that adds optional static typing and that adopts the latest ECMAScript features before they are broadly deployed by browser vendors. It gets transpiled to JavaScript for deployment, and the support for debugging using sourcemaps is currently good.

The following steps need to happen:
- setup build system to use TypeScript and Rollup to bundle the (JavaScript) sources
- convert JavaScript files to TypeScript by adding type annotation, particularly for arguments and return values
- correct any errors that will be discovered as part of this process (mistyped property names is a common one)
- switch the test harness from `tape` to `jest`

Once completed, the codebase will be easier to maintain and the MathLive interface will have a "native" TypeScript interface (instead of the current one which is generated from JSDoc annotations and is not as expressive as desirable as a result). This will make the integration of MathLive in TypeScript projects easier. On the other hand, this should be transparent for JavaScript projects using MathLive, as the regular JavaScript API will still be available.

**Comment by arnog** [comment]
Update:
- the build system is now setup
- the test harness and the test cases have been migrated to `jest`
- the core rendering engine (about half the codebase) has been converted

Left to do:
- convert the editor (about half the codebase)
- run the linter with more stringent options (`explicit-function-return-type`, `prefer-optional-chain`, `prefer-nullish-coalescing`, `explicit-function-return-type`)
- review the generated type info for the main API

**Comment by arnog** [comment]
The migration of the editor is in progress. More bugs found, for example:
```
if (options && options.onUndoStateWillChange === 'function') {
    options.onUndoStateWillChange();
}
```
Perfectly valid JavaScript, although semantically incorrect. There is a `typeof` missing before `options.onUndoStateWillChange`. As a result, the `onUndoStateWillChange()` handler was never called, but Typescript noticed!

As part of the migration, major refactoring of the codebase. The editor currently includes two gigantic monolithic files:
- `editor-mathlist.js`: 3,922 lines
- `editor-mathfield.js`: **4,155 lines. Yikes.**

They are being split apart and organized into more granular components. And `mathlist` is also being renamed to `model` which better describes its functionality and is more consistent with the terminology of other editors (monaco, code-mirror, etc...):
- `editor-mathlist.js` → 
   - `model.ts`
   - `model-utils.ts`
   - `model-styling.ts`
   - `model-smartfence.ts`
   - `model-selection.ts`
   -  `model-listeners.ts`
   - `model-insert.ts`
   - `model-delete.ts`
   - `model-commands.ts`
   - `model-command-mode.ts`
   - `model-array.ts`
   - `model-array-utils.ts`
- `editor-mathfield.js` → 
   - `a11y.ts`
   - `autocomplete.ts`
   - `commands.ts`
   - `config.ts`
   - `speech.ts`
   - `speech-read-aloud.ts`
   - `undo.ts`
   - `mathfield.ts`
   - `mathfield-virtual-keyboards.ts`
   - `mathfield-utils.ts`
   - `mathfield-styling.ts`
   - `mathfield-smartmode.ts`
   - `mathfield-render.ts`
   - `mathfield-pointer-input.ts`
   - `mathfield-keyboard-input.ts`
   - `mathfield-commands.ts`
   - `mathfield-clipboard.ts`
   - `mathfield-buttons.ts`

Some functionality that was entangled a bit all over the place (speech support, keyboard input, configuration) is now regrouped into separate modules. In addition, there's a brand new system for dealing with commands (i.e. things that can be invoked with `perform()`) that makes them more explicit and data driven (previously commands were just methods on certain classes that had a special naming convention).

All of those changes are internal and do not affect the public API. Note there are plans to alter the API as documented in #159 and #160 and more, but this will be the subject of a future tracked issue.

A question for anyone using MathLive with TypeScript: would you prefer to have a single declaration file, i.e. `mathlive.d.ts`, or several, i.e. `mathlive.d.ts`, `mathfield.d.ts` and `mathfied-config.d.ts`?

**Comment by leerobert** [comment]
> A question for anyone using MathLive with TypeScript: would you prefer to have a single declaration file, i.e. mathlive.d.ts, or several, i.e. mathlive.d.ts, mathfield.d.ts and mathfied-config.d.ts?

Hm it might make sense to have those concepts broken up into their respective declarative files. Would there be a savings in speed/overhead on import? I'm not enough of an expert in js/ts to speculate.

**Comment by arnog** [comment]
The performance impact would be negligible. Remember that these `.d.ts` files are only used locally, they are not downloaded over the wire. So, it's really a question of what would be more clear and easier to use for the developer. I'm still on the fence...

**Comment by arnog** [comment]
I've completed:
- migration of the editor
- separation of public and private API, and creation of `.d.ts` files for the public API

Still left to do before pushing to MASTER:
- additional testing, in particular of areas where bugs were discovered. I need to verify that the fixes are correct.
- writing a documentation generation tool

The documentation was previously generated with `jsdoc`. The TypeScript declaration files (`.d.ts`) will now be the source of truth for the documentation. But I couldn't find a tool that generate the documentation in the format I want, so I will need to write one. It will use `tsdoc` to parse and generate an AST of the declaration files, then I will parse the JSON and generate markup as needed.

I need to complete this step, because otherwise once I push to MASTER, `docs.mathlive.io` will get updated with some blank documentation...

**Comment by arnog** [comment]
I've been making good progress, but it has taken longer than I hoped. I've written a documentation generation tool that will take the TypeScript declaration files as input and generate a documentation page. It turned out to be a bit more involved than I hoped (and I looked around for existing tools to do this, but they did not do quite what I needed).

You can [preview it here](http://cortexjs.io/docs/mathlive/). I would love to hear feedback, especially compared to the ["old" documentation](https://docs.mathlive.io).

A few things to notice:
- the documentation is much more comprehensive. It includes the commands for `perform()` and all the configuration options (which were in separate handwritten pages before, now they are generated from code and therefore guaranteed to be more accurate and up to date).
- Some things have been broken up, for example the config object is made up of an intersection of categories of options (Editing, Speech, Listeners, etc...)
- all the references are linked, so if you need more info about `InlineShortcutDefinition` while reading about `InlineShortcutsOptions`, just click on the link. This turned out to be "interesting" to implement (did you know that in a class you could have a constructor, an instance method named `constructor` and a static method named `constructor`? Fun times). You can also get URLs to specific entries, for example [`makeMathField()`](http://cortexjs.io/docs/mathlive/#(%22mathlive%22%3Anamespace).(makeMathField%3Afunction)).

I still need to do an editorial pass on the documentation (some types are still missing, some comments need to be added, some links to external resources are broken) but I'm interested in feedback on the legibility of the typography and layout of the info, the ease of navigation, etc...

**Comment by physedo** [comment]
Very Good

**Comment by NSoiffer** [comment]
It looks pretty good, but `options: any` is not helpful. It would be much more helpful to have a comment to explain the options or maybe better from a code correctness point of view, eliminate "any" and be specific about what the options really are from a type point of view.

Similarly, "string" is not helpful -- what values can the string take on?

I did spot a bug dealing with links on http://cortexjs.io/docs/mathlive/#(%22core%22%3Anamespace).(ParseMode%3Atype):  

> textToSpeechRulesOptions?: {[key: string]: string};A set of key/value pairs that can be used to configure the speech rule engine.Which options are available depends on the speech rule engine in use. There are no options available with MathLive’s built-in engine. The options for the SRE engine are documented {@link:https://github.com/zorkow/speech-rule-engine | here}

One the webpage, that link is just plain text, not a clickable link.

**Comment by arnog** [comment]
Yes, the missing types are part of the editorial pass I need to make. I have the proper types on the implementation side, but I need to surface them on the declaration side. There shouldn’t be any `any` left, and strings that can only take some values will say so explicitly.

Good catch on the `@link` issue. I’m using a stricter syntax, and I need to revise the tag to match it.

**Comment by arnog** [comment]
OK, I've pushed the TypeScript Migration to master. I hope things are not too broken.

The documentation has migrated to https://cortexjs.io/docs/mathlive/ (docs.mathlive.io will redirect there).

I've published the documentation generator as a separate project: https://github.com/ui-js/grok

I ended up going with multiple TypeScript declaration files (I couldn't get a single file to work). In the vast majority of cases, you would still need to only import the top-level public module, though: 'mathlive' (-> 'mathlive.d.ts').
